persistence/sql/persister_recovery.go

```diff
@@ -74,7 +74,7 @@ func (p *Persister) UseRecoveryToken(ctx context.Context, token string) (*link.R
 	nid := p.NetworkID(ctx)
 	if err := sqlcon.HandleError(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) (err error) {
 		for _, secret := range p.r.Config().SecretsSession(ctx) {
-			if err = tx.Where("token = ? AND nid = ? AND NOT used", p.hmacValueWithSecret(ctx, token, secret), nid).First(&rt); err != nil {
+			if err = tx.Where("token = ? AND nid = ? AND NOT used AND selfservice_recovery_flow_id = ?", p.hmacValueWithSecret(ctx, token, secret), nid, fID).First(&rt); err != nil {
```

Comment:
Is `selfservice_recovery_flow_id` nullable? If so, this query here will never be `selfservice_recovery_flow_id = NULL`.

Reply:
It is nullable and `selfservice_recovery_flow_id = NULL` can be true for older flows, as we can't infer the flow ID in a migration. Though with this implementation the parameter `fID` will _never_ be null, as there _is_ always a flow. In case of an admin initiated flow, it was just not set in the `selfservice_recovery_flow_id` column.

Which means these flows would no longer work with this change. I guess, I could add a check to see if the flow is an `Admin` flow, and if so, not pass a `fID` in, but that would kinda defeat the purpose of this change, IMO.

I am not sure how to handle these "legacy" flows.

Reply:
This would mean that all admin recovery flows initiated before an upgrade and completed after the upgrade would fail. I think that is not pretty, but acceptable, given that admin recovery flows are not that frequent and the admin knows about the upgrade. Can you please add in the PR message body a `BREAKING CHANGE` note to inform users about this change and impact?

Reply:
Done.